### ISSUE TITLE
startScript doesn't end on Bun and Deno

### ISSUE THREAD

**Issue by wellwelwel** [wellwelwel/poku#143]
The `startScript` currently doesn't work for **Bun** and **Deno**.

### How the `startScript` works?

- It will use `spawn` to start your script in a background process
- By using `.end()`, the current process and all its sub-processes will be terminated
- By forcing finish (`Ctrl+C`), it will terminate all the background process and all its sub-processes

### Expected Behavior

Using **Node.js**, it will close all subprocesses when closing the main process.

### The Issue

However, on **Bun** and **Deno**, when the main process is finished, all subprocesses remain open and only the main process is closed.

---

**PR**s are welcome 🤝

**Comment by wellwelwel** [comment]
Same issue on Windows was fixed in #118 🎉

**Comment by hiroshimorowaka** [comment]
Boa tarde! Eu gostaria de contribuir nessa Issue, mas eu não to conseguindo subir os containers de docker pra rodar os testes. Tem algo estou fazendo de errado?

**Comment by wellwelwel** [comment]
> Good afternoon! I'd like to contribute to this Issue, but I can't get the docker containers to run the tests. Is there something I'm doing wrong?
> And how can I test the startScript and see the behavior of the Child process in bun, deno and node?

To test using the containers, you will need to perform `npm run test:bun`, `npm run test:deno` or `npm run test:node`.

- I'm not sure about this on Windows, since it uses local environment variables.

Docker commands are called using `docker compose up` ([source](https://github.com/wellwelwel/poku/blob/12bf513299407325b1af38eda2533dcb29481952/test/docker/docker-compose.yml)), and you can find all `Dockerfiles` [here](https://github.com/wellwelwel/poku/tree/12bf513299407325b1af38eda2533dcb29481952/test/docker).

- [This helper](https://github.com/wellwelwel/poku/blob/12bf513299407325b1af38eda2533dcb29481952/test/helpers/check-node.test.ts) is who calls the `docker compose` commands.
- [`startScript` and `startService` source](https://github.com/wellwelwel/poku/blob/12bf513299407325b1af38eda2533dcb29481952/src/modules/create-service.ts).

---

An unconventional way is installing **Poku** in a test project using your local fork, instead of `npm` directly. For example:

```bash
# in your fork path
npm run build
pwd
```

Then, you can copy the result from `pwd` command and in your test directory path:

```bash
npm i -D PWD_RESULT
```

That way, every time you run `npm run build` on your **Poku** fork, it will be updated in your test path directly.
You can create a simple server and test it (see some basic servers examples in the [examples](https://poku.io/docs/examples/local-server)) 🙋🏻‍♂️

---

<details>
<summary>
<b>Português (BR) 🇧🇷</b>
</summary>

> Boa tarde! Eu gostaria de contribuir nessa Issue, mas eu não to conseguindo subir os containers de docker pra rodar os testes. Tem algo estou fazendo de errado?
> E como eu posso testar o startScript e ver o comportamento do Child process no bun, deno e node ?

Eaí @hiroshimorowaka, ao executar `npm run test`, isso irá executar todos os testes localmente (sem necessidade de containers).

Para testar usando containers, você precisa executar `npm run test:bun`, `npm run test:deno` or `npm run test:node`.

- Não tenha certeza quanto a isso no Windows, pois esses comandos usam variáveis locais.

Os comandos do Docker são chamados usando `docker compose up` ([origem](https://github.com/wellwelwel/poku/blob/12bf513299407325b1af38eda2533dcb29481952/test/docker/docker-compose.yml)), e você pode encontrar todos os `Dockerfiles` [aqui](https://github.com/wellwelwel/poku/tree/12bf513299407325b1af38eda2533dcb29481952/test/docker).

- [Esse auxiliar](https://github.com/wellwelwel/poku/blob/12bf513299407325b1af38eda2533dcb29481952/test/helpers/check-node.test.ts) é quem chama os comandos `docker compose`.
- [Código fonte dos `startScript` e `startService`](https://github.com/wellwelwel/poku/blob/12bf513299407325b1af38eda2533dcb29481952/src/modules/create-service.ts).

---

Uma forma não tão convencional é instalando o **Poku** em um diretório de teste usando seu próprio _fork_, ao invés do **npm** diretamente. Por exemplo:

```bash
# No caminho do seu fork
npm run build
pwd
```

Então, você pode copiar o resultado retornado do `pwd` e testar diretamente no seu diretório com o seu teste:

```bash
npm i -D PRESULTADO_DO_PWD
```

Dessa forma, toda vez que você executar `npm run build` no seu _fork_, seu diretório de teste será atualizado junto.

Você pode criar um servidor super simples e testar (veja alguns exemplos básicos na seção de [exemplos](https://poku.io/docs/examples/local-server)).

</details>

**Comment by hiroshimorowaka** [comment]
My suspicion about the problem is as follows: 
NodeJS, Bun and Deno's Spawn API under the hood work in different ways, which is why NodeJS's `process.kill` doesn't work on Deno or Bun. 
I was reading [Bun's documentation](https://bun.sh/docs/api/spawn#benchmarks) and it seems that it works based on [posix_spawn](https://man7.org/linux/man-pages/man3/posix_spawn.3.html). Both Deno and Bun have their own methods of creating and removing child processes. 

[Bun compatibility](https://bun.sh/docs/runtime/nodejs-apis) with NodeJS Child Process is partial,

I don't know if there is a way to integrate Bun and Deno API's in this specific case, I imagine not. 

I was thinking of maybe using [exec](https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback) instead of `spawn`, but I don't know exactly how it works and if it would solve this problem.

**Comment by wellwelwel** [comment]
**Bun** also has it own [`spawn`](https://bun.sh/guides/process/spawn) and the same issue occurs on it 😕

A workaround could be suggest Bun and Deno users to close your ports using [**kill-port**](https://www.npmjs.com/package/kill-port) for **Bun** and **Deno** and remove the intentinal `throw`:

https://github.com/wellwelwel/poku/blob/12bf513299407325b1af38eda2533dcb29481952/src/modules/create-service.ts#L186-L189

---

This way won't fix the issue, but will allow **Bun** and **Deno** users to use this feature.

**Comment by wellwelwel** [comment]
> which is why NodeJS's process.kill doesn't work on Deno or Bun.

The main process is closed correctly, the issue occurs with the subprocesses (different **PID**s).

Both **Bun** and **Deno** will close only the main process (specified **PID**) and keeping subprocesses open.

**Comment by wellwelwel** [comment]
> I was thinking of maybe using [exec](https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback) instead of spawn, but I don't know exactly how it works and if it would solve this problem.

You can consider as example the `startService`, that will work for both **Node.js**, **Bun** and **Deno** and is covered with tests.

It works because the `startService` starts a single process, then, it will close the **PID** correctly.

**Comment by wellwelwel** [comment]
Fixed for **Deno** in #242 by passing a port to `end` method.

---

For **Bun**, there are two issues that need to be fixed on the **Bun** side:

- To use `.end` without pass a port:
  - https://github.com/oven-sh/bun/issues/7441
- To use `.end` passing a port:
  - https://github.com/oven-sh/bun/issues/11055

---

Even though it's not yet possible to use this feature with **Bun**, I'll close it as complete because the errors need to be fixed directly in **Bun**.

When one or both bugs are fixed in **Bun**, it will automatically be possible to use this feature.

**Comment by wellwelwel** [comment]
> _Nevermind_

Fixed for **Bun** in #245 by passing a port to end method.
Fixed for **Deno** in #242 by passing a port to end method.
